File: app/src/test/java/com/leaguechampions/ui/championdetails/ChampionDetailsViewModelTest.java
package com.leaguechampions.ui.championdetails;

import android.arch.core.executor.testing.InstantTaskExecutorRule;
import android.arch.lifecycle.Observer;

import com.leaguechampions.data.repository.Resource;
import com.leaguechampions.data.model.RiotResponse;
import com.leaguechampions.data.repository.ChampionRepository;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.reset;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

@RunWith(MockitoJUnitRunner.class)
public class ChampionDetailsViewModelTest {

    @Rule public InstantTaskExecutorRule instantExecutorRule = new InstantTaskExecutorRule();

    @Mock ChampionRepository championRepository;
    @Mock Observer<Resource<RiotResponse>> observer;

    @Captor ArgumentCaptor<String> stringCaptor;

    private ChampionDetailsViewModel viewModel;
    private String championId = "Riven";

    @Before
    public void setUp() {
        viewModel = new ChampionDetailsViewModel(championRepository);
    }

    @After
    public void tearDown() {

    }

    @Test
    public void testSetChampionId_WhenChampionIdIsNull_DoNothing() {
        viewModel.getRiotResponse().observeForever(observer);

        viewModel.setChampionId(null);

        verify(championRepository, never()).getChampionDetails(anyString());
    }

    @Test
    public void testSetChampionId_WhenChampionIdIsEqual_GetChampionDetailsOnlyOnce() {
        viewModel.getRiotResponse().observeForever(observer);

        viewModel.setChampionId(championId);
        viewModel.setChampionId(championId);

        verify(championRepository, times(1)).getChampionDetails(stringCaptor.capture());
        assertThat(stringCaptor.getValue(), is(championId));
    }

    @Test
    public void testSetChampionId_WhenChampionIdIsDifferent_GetChampionDetails() {
        viewModel.getRiotResponse().observeForever(observer);

        viewModel.setChampionId(championId);

        verify(championRepository).getChampionDetails(stringCaptor.capture());
        assertThat(stringCaptor.getValue(), is(championId));

        reset(championRepository);

        viewModel.setChampionId("Jinx");

        verify(championRepository).getChampionDetails(stringCaptor.capture());
        assertThat(stringCaptor.getValue(), is("Jinx"));
    }
}

File: app/src/test/java/com/leaguechampions/ui/championdetails/ChampionDetailsViewModelTest.kt
package com.leaguechampions.ui.championdetails

import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import com.leaguechampions.core.data.model.Champion
import com.leaguechampions.core.data.repository.ChampionRepository
import com.leaguechampions.core.utils.Resource
import com.leaguechampions.utils.Status
import org.assertj.core.api.Assertions.assertThat
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TestRule
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.junit.MockitoJUnitRunner

@RunWith(MockitoJUnitRunner::class)
class ChampionDetailsViewModelTest {

    private val championId = "Riven"
    private val champion = com.leaguechampions.core.data.model.Champion(championId, "", "", "", "")
    private val error = "error"
    private val viewStateLoading = com.leaguechampions.features.champions.championdetails.ChampionDetailsViewModel.ViewState(status = Status.LOADING)
    private val viewStateSuccess = com.leaguechampions.features.champions.championdetails.ChampionDetailsViewModel.ViewState(status = Status.SUCCESS, champion = champion)
    private val viewStateError = com.leaguechampions.features.champions.championdetails.ChampionDetailsViewModel.ViewState(status = Status.ERROR, error = error)

    private lateinit var viewModel: com.leaguechampions.features.champions.championdetails.ChampionDetailsViewModel

    @Mock private lateinit var repository: com.leaguechampions.core.data.repository.ChampionRepository
    @Mock private lateinit var observerViewState: Observer<com.leaguechampions.features.champions.championdetails.ChampionDetailsViewModel.ViewState>

    @get:Rule var rule: TestRule = InstantTaskExecutorRule()

    @Before
    fun setUp() {
        viewModel = com.leaguechampions.features.champions.championdetails.ChampionDetailsViewModel(repository)
        viewModel.viewState.observeForever(observerViewState)
    }

    @After
    fun tearDown() {

    }

    @Test
    fun testSetChampionId_loading() {
        val resource = MutableLiveData<com.leaguechampions.core.utils.Resource<com.leaguechampions.core.data.model.Champion>>()
        resource.value = com.leaguechampions.core.utils.Resource.loading()
        `when`(repository.getChampionDetails(championId)).thenReturn(resource)

        viewModel.setChampionId(championId)

        verify(repository).getChampionDetails(championId)
        assertThat(viewModel.viewState.value).isEqualTo(viewStateLoading)
        verifyNoMoreInteractions(repository)
    }

    @Test
    fun testSetChampionId_success() {
        val resource = MutableLiveData<com.leaguechampions.core.utils.Resource<com.leaguechampions.core.data.model.Champion>>()
        resource.value = com.leaguechampions.core.utils.Resource.success(champion)
        `when`(repository.getChampionDetails(championId)).thenReturn(resource)

        viewModel.setChampionId(championId)

        verify(repository).getChampionDetails(championId)
        assertThat(viewModel.viewState.value).isEqualTo(viewStateSuccess)
        verifyNoMoreInteractions(repository)
    }

    @Test
    fun testSetChampionId_error() {
        val resource = MutableLiveData<com.leaguechampions.core.utils.Resource<com.leaguechampions.core.data.model.Champion>>()
        resource.value = com.leaguechampions.core.utils.Resource.error(error)
        `when`(repository.getChampionDetails(championId)).thenReturn(resource)

        viewModel.setChampionId(championId)

        verify(repository).getChampionDetails(championId)
        assertThat(viewModel.viewState.value).isEqualTo(viewStateError)
        verifyNoMoreInteractions(repository)
    }

    @Test
    fun testSetChampionId_null() {
        viewModel.setChampionId(null)

        verify(repository, never()).getChampionDetails(anyString())
        assertThat(viewModel.viewState.value).isEqualTo(null)
        verifyNoMoreInteractions(repository)
    }
}
